src/editor.ts

```diff
@@ -0,0 +1,675 @@
+                    scaleMax = Math.max(scale_0[i], scale_1[i], scale_2[i]);
+                } else {
+                    scaleMin = Math.min(scaleMin, scale_0[i], scale_1[i], scale_2[i]);
+                    scaleMax = Math.max(scaleMax, scale_0[i], scale_1[i], scale_2[i]);
+                }
+            }
+
+            const maxScale = Math.log(Math.exp(scaleMin) + value * (Math.exp(scaleMax) - Math.exp(scaleMin)));
+
+            processSelection(state, op, (i) => scale_0[i] > maxScale || scale_1[i] > maxScale || scale_2[i] > maxScale);
+        });
+        updateState();
+    });
+
+    events.on('select.byOpacity', (op: string, value: number) => {
+        splatDefs.forEach((splatDef) => {
+            const splatData = splatDef.data;
+            const state = splatData.getProp('state') as Uint8Array;
+            const opacity = splatData.getProp('opacity') as Float32Array;
+
+            processSelection(state, op, (i) => {
+                const t = Math.exp(opacity[i]);
+                return ((1 / (1 + t)) < value);
+            });
+        });
+        updateState();
+    });
+
+    events.on('select.bySpherePlacement', (sphere: number[]) => {
+        debugSphereCenter.set(sphere[0], sphere[1], sphere[2]);
+        debugSphereRadius = sphere[3];
+
+        scene.forceRender = true;
+    });
+
+    events.on('select.byPlanePlacement', (axis: number[], distance: number) => {
+        debugPlane.set(axis[0], axis[1], axis[2]);
+        debugPlaneDistance = distance;
+
+        scene.forceRender = true;
+    });
+
+    events.on('select.bySphere', (op: string, sphere: number[]) => {
+        splatDefs.forEach((splatDef) => {
+            const splatData = splatDef.data;
+            const state = splatData.getProp('state') as Uint8Array;
+            const x = splatData.getProp('x');
+            const y = splatData.getProp('y');
+            const z = splatData.getProp('z');
+
+            const radius2 = sphere[3] * sphere[3];
+            vec.set(sphere[0], sphere[1], sphere[2]);
+
+            mat.invert(splatDef.element.worldTransform);
+            mat.transformPoint(vec, vec);
+
+            processSelection(state, op, (i) => {
+                vec2.set(x[i], y[i], z[i]);
+                return vec2.sub(vec).lengthSq() < radius2;
+            });
+        });
+        updateState();
+    });
+
+    events.on('select.byPlane', (op: string, axis: number[], distance: number) => {
+        splatDefs.forEach((splatDef) => {
+            const splatData = splatDef.data;
+            const state = splatData.getProp('state') as Uint8Array;
+            const x = splatData.getProp('x');
+            const y = splatData.getProp('y');
+            const z = splatData.getProp('z');
+
+            vec.set(axis[0], axis[1], axis[2]);
+            vec2.set(axis[0] * distance, axis[1] * distance, axis[2] * distance);
+
+            // transform the plane to local space
+            mat.invert(splatDef.element.worldTransform);
```

Comment:
Not entirely sure what matrix gets inverted here .. just making sure this needs to take place in the loop

Reply:
The loop here is over instances of GSplat in the scene, not individual splats. The loop over splats happens in `processSelection` below. 👍🏻